### PR TITLE
Login

### PR DESCRIPTION
**Pull request by at-dinhtruong**
#1

### CODE REVIEW

**Comment by at-dinhtruong** [comment]
#1

File: app/src/main/java/vn/asiantech/travelmate/login/LoginActivity.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+class LoginActivity : AppCompatActivity() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_login)
+        initFragment()
+    }
+
+    private fun initFragment() {
+        supportFragmentManager.beginTransaction().apply {
```
Comment (at-tainguyen) [inline app/src/main/java/vn/asiantech/travelmate/login/LoginActivity.kt]: In this case, you should implement like below 
```
supportFragmentManager
            .beginTransaction()
            .replace(R.id.fragment_container, LoginFragment())
            .commit()
```
It's make the code is clearer

File: app/src/main/java/vn/asiantech/travelmate/login/LoginFragment.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import vn.asiantech.travelmate.utils.Validate
+
+class LoginFragment : Fragment(), View.OnClickListener {
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        val view: View = inflater.inflate(R.layout.fragment_login, container, false)
+        initView(view)
+        return view
+    }
+
+    private fun initView(view: View?) {
```
Comment (at-tainguyen) [inline app/src/main/java/vn/asiantech/travelmate/login/LoginFragment.kt]: Do not need `view` here. You can directly access to `tvRegister` and `btnLogin` without `view`

File: app/src/main/java/vn/asiantech/travelmate/login/SignUpFragment.kt [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+class SignUpFragment : Fragment(), View.OnClickListener {
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        val view: View = inflater.inflate(R.layout.fragment_sign_up, container, false)
+        initView(view)
+        return view
+    }
+
+    private fun initView(view: View?) {
```
Comment (at-tainguyen) [inline app/src/main/java/vn/asiantech/travelmate/login/SignUpFragment.kt]: Do not need `view` here. You can directly access to `tvRegister` and `btnLogin` without `view`

File: app/src/main/java/vn/asiantech/travelmate/login/SignUpFragment.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        view?.tvLogin?.setOnClickListener(this)
+        view?.btnSignUp?.setOnClickListener(this)
+    }
+
+    override fun onClick(v: View?) {
+        if (v?.id == R.id.btnSignUp) {
+            if (checkUserPassEmail() == "pass") {
```
Comment (at-tainguyen) [inline app/src/main/java/vn/asiantech/travelmate/login/SignUpFragment.kt]: Please create the `Mock` file to hold all mock data. 
Avoid hard-code

File: app/src/main/java/vn/asiantech/travelmate/utils/Validate.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package vn.asiantech.travelmate.utils
+
+import java.util.regex.Pattern
+
+class Validate {
```
Comment (at-tainguyen) [inline app/src/main/java/vn/asiantech/travelmate/utils/Validate.kt]: ValidationUtil should like below 
```
object ValidationUtil {
    private const val PASSWORD_PATTERN = "^(?=.*[0-9])(?=.*[a-zA-Z]).{6,18}$"

    fun isValidPassword(password: String): Boolean = Pattern.compile(PASSWORD_PATTERN).matcher(password).matches()
}
```
and use like below 
```
ValidationUtil.isValidPassword(password)
```
is correct way

File: app/src/main/java/vn/asiantech/travelmate/login/SignUpFragment.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    private fun checkUserPassEmail(): String {
+        val validate = Validate()
+        val firstName = edtFirstName.text.toString().trim()
+        val lastName = edtLastName.text.toString().trim()
+        val email = edtEmail.text.toString().trim()
+        val password = edtPassword.text.toString().trim()
+        val confirmPassword = edtConfirmPassword.text.toString().trim()
+        return if (!validate.isValidFirstName(firstName)) {
```
Comment (at-tainguyen) [inline app/src/main/java/vn/asiantech/travelmate/login/SignUpFragment.kt]: This block should convert to `when` like below 
```
return when {
            password != confirmPassword -> getString(R.string.signupTvConfirmPasswordWrong)
            !validate.isValidEmail(email) -> getString(R.string.emailFormatWrong)
            !validate.isValidFirstName(firstName) -> getString(R.string.signupTvFirstNameFormatWrong)
            !validate.isValidLastName(lastName) -> getString(R.string.signupTvLastNameFormatWrong)
            !validate.isValidPassword(password) -> getString(R.string.signupTvPasswordFormatWrong)
            else -> "pass"
        }
``` to make code clearer

File: app/src/main/java/vn/asiantech/travelmate/login/SignUpFragment.kt [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            "pass"
+        }
+    }
+
+    private fun showMessage(message: String) {
+        val slideUp = AnimationUtils.loadAnimation(context, R.anim.slide_up)
+        val slideDown = AnimationUtils.loadAnimation(context, R.anim.slide_down)
+        val handler = Handler()
+        if (tvMessageError.visibility == View.INVISIBLE) {
+            tvMessageError.visibility = View.VISIBLE
```
Comment (at-tainguyen) [inline app/src/main/java/vn/asiantech/travelmate/login/SignUpFragment.kt]: Should using `apply` for `tvMessageError` in this case

File: app/src/main/java/vn/asiantech/travelmate/login/LoginFragment.kt [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    override fun onClick(v: View?) {
+        if (v?.id == R.id.tvRegister) {
+            fragmentManager?.beginTransaction()?.apply {
+                setCustomAnimations(R.anim.right_to_left1, R.anim.right_to_left2, R.anim.left_to_right1, R.anim.left_to_right2)
+                add(R.id.fragment_container, SignUpFragment())
+                addToBackStack(null)
+                commit()
+            }
+        } else {
+            if (ValidationUtil.isValidEmail(email = edtEmail?.text.toString().trim()) && ValidationUtil.isValidPassword(password = edtPassword?.text.toString().trim())) {
```
Comment (at-tainguyen) [inline app/src/main/java/vn/asiantech/travelmate/login/LoginFragment.kt]: Do not need named argument `email`, `password` here.
In this case, the source code should like below : 
```
if (ValidationUtil.isValidEmail(edtEmail?.text.toString().trim()) && ValidationUtil.isValidPassword(edtPassword?.text.toString().trim())) {
.....
}
```

Please check this document about `Kotlin Default and Named Arguments`

File: app/src/main/java/vn/asiantech/travelmate/login/LoginFragment.kt [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+                Toast.makeText(context, "Successful", Toast.LENGTH_SHORT).show()
```
Comment (at-tainguyen) [inline app/src/main/java/vn/asiantech/travelmate/login/LoginFragment.kt]: Please do not forget avoid hard-code

**Pull request merged**